### PR TITLE
Use notifications to manage parameter changes

### PR DESCRIPTION
**Pull request by AAJELLAL**

### CODE REVIEW

File: src/components/computing-status/computing-type.ts [hunk 1]
```diff
@@ -17,4 +17,8 @@ export enum ComputingType {
     STATE_ESTIMATION = 'STATE_ESTIMATION',
 }
 
+export const isValidComputingType = (value: string): boolean => {
```
Comment (sBouzols) [inline src/components/computing-status/computing-type.ts:20]: Maybe rename this `ComputationType` instead of `ComputingType` like in the backend...

File: src/components/study-container.jsx [hunk 2]
```diff
@@ -313,10 +316,18 @@ export function StudyContainer({ view, onChangeTab }) {
...
                 if (updateTypeHeader === 'STUDY_ALERT') {
                     sendAlert(eventData);
                     return; // here, we do not want to update the redux state
                 }
+                // used to update the computation type
+                if (
+                    updateTypeHeader === NotificationType.COMPUTATION_PARAMETERS &&
+                    isValidComputingType(computationTypeHeader)
+                ) {
+                    dispatch(setStudyParamsChanged(computationTypeHeader));
```
Comment (sBouzols) [inline src/components/study-container.jsx:329]: Instead of adding a new entry in redux store, you should use the existing one I think :
`dispatch(studyUpdated(eventData));`
One Websocket, one state in the store, multiple consumers which filter interesting notifications by HEADERS values.

File: src/components/dialogs/parameters/parameters.jsx [hunk 3]
```diff
@@ -437,6 +431,32 @@ export const useParametersBackend = (
         }
     }, [optionalServiceStatus, backendFetchParameters, snackError, studyUuid, type]);
 
+    // we need to call backendFetchParameters when ever the study params change
+    useEffect(() => {
```
Comment (sBouzols) [inline src/components/dialogs/parameters/parameters.jsx:435]: This is almost perfectly the same effect as the previous one.
Avoid code duplication.

File: src/components/dialogs/parameters/sensi/sensitivity-analysis-parameters.jsx [hunk 4]
```diff
@@ -454,6 +454,11 @@ export const SensitivityAnalysisParameters = ({ parametersBackend, setHaveDirtyF
         isSubmitAction,
         reset,
     ]);
+    useEffect(() => {
+        if (params) {
+            reset(fromSensitivityAnalysisParamsDataToFormValues(params));
+        }
+    }, [params, reset, fromSensitivityAnalysisParamsDataToFormValues]);
```
Comment (sBouzols) [inline src/components/dialogs/parameters/sensi/sensitivity-analysis-parameters.jsx:461]: in this file :
```
        setSensitivityAnalysisParameters(studyUuid, null)
            .then(() => {
            .then(() => {
                return getSensitivityAnalysisParameters(studyUuid)
                return getSensitivityAnalysisParameters(studyUuid)
                    .then((params) => {
                    .then((params) => {
                        setSensitivityAnalysisParams(params);
                        setSensitivityAnalysisParams(params);
                        updateParameters(params);
                        updateParameters(params);
                    })
                    })
                    .catch((error) => {
                    .catch((error) => {
                        snackError({
                        snackError({
                            messageTxt: error.message,
                            messageTxt: error.message,
                            headerId: 'paramsRetrievingError',
                            headerId: 'paramsRetrievingError',
                        });
                        });
                    });
                    });
            })
            })
            .catch((error) => {
            .catch((error) => {
                snackError({
                snackError({
                    messageTxt: error.message,
                    messageTxt: error.message,
                    headerId: 'paramsChangingError',
                    headerId: 'paramsChangingError',
                });
                });
            });
            });
 ```
 You don't need to get after set if notification trigger an update.
 You don't need a local `sensitivityAnalysisParams` if you already have a synchronysed one (`params`)
 
 This file needs to be clean

File: src/components/dialogs/parameters/short-circuit-parameters.jsx [hunk 5]
```diff
@@ -65,6 +71,26 @@ export const useGetShortCircuitParameters = () => {
...
+                .then((params) => {
+                    setShortCircuitParams(params);
+                    dispatch(setStudyParamsChanged(''));
+                })
+                .catch((error) => {
+                    snackError({
+                        messageTxt: error.message,
+                        headerId: 'paramsRetrievingError',
+                    });
+                });
```
Comment (sBouzols) [inline src/components/dialogs/parameters/short-circuit-parameters.jsx:90]: same code duplication...

I understand that we want to fetchparams by notification OR when studyUuid OR shortCircuitAvailability changed. We could write multiple effect to defines those cases, but the consequence should not be duplicated.

File: src/components/dialogs/parameters/voltageinit/use-get-voltage-init-parameters.tsx [hunk 6]
```diff
@@ -39,5 +43,25 @@ export const useGetVoltageInitParameters = (): [
         }
     }, [voltageInitAvailability, studyUuid, snackError]);
 
+    useEffect(() => {
+        if (
+            studyUuid &&
+            voltageInitAvailability === OptionalServicesStatus.Up &&
+            studyParamsChanged === ComputingType.VOLTAGE_INITIALIZATION
+        ) {
+            getVoltageInitStudyParameters(studyUuid)
```
Comment (sBouzols) [inline src/components/dialogs/parameters/voltageinit/use-get-voltage-init-parameters.tsx:52]: same code duplication

File: src/components/dialogs/parameters/widget/parameter-line-slider.tsx [hunk 7]
```diff
@@ -32,6 +32,12 @@ const ParameterLineSlider = ({
     const [parameterValue, handleChangeParameterValue] = useParameterState(paramNameId);
     const [sliderValue, setSliderValue] = useState(Number(parameterValue));
 
+    useEffect(() => {
+        if (parameterValue) {
+            setSliderValue(Number(parameterValue));
+        }
+    }, [parameterValue]);
```
Comment (sBouzols) [inline src/components/dialogs/parameters/widget/parameter-line-slider.tsx:39]: This is a fix for another scope ? I mean it's wasn't working before your development...
Maybe do it in another PR ?

File: src/components/map-viewer.jsx [hunk 8]
```diff
@@ -5,7 +5,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
 
-import { useState, useCallback, useRef, useEffect } from 'react';
+import React, { useState, useCallback, useRef, useEffect } from 'react';
```
Comment (sBouzols) [inline src/components/map-viewer.jsx:8]: remove this change

File: src/redux/reducer.ts [hunk 9]
```diff
@@ -297,6 +301,7 @@ export interface StudyUpdatedEventDataHeader {
     nodes?: UUID[];
     error?: string;
     userId?: string;
+    paramsName?: string;
```
Comment (sBouzols) [inline src/redux/reducer.ts:304]: you mean `computationType` here ?

File: src/redux/reducer.ts [hunk 10]
```diff
@@ -274,6 +277,7 @@ import { StudyDisplayMode } from '../components/network-modification.type';
 
 export enum NotificationType {
     STUDY = 'study',
+    COMPUTATION_PARAMETERS = 'computationParametersUpdated',
```
Comment (sBouzols) [inline src/redux/reducer.ts:280]:
```suggestion
    COMPUTATION_PARAMETERS_UPDATED = 'computationParametersUpdated',
```

File: src/components/study-container.jsx (hunk 2 above)
Comment (AAJELLAL) [reply src/components/study-container.jsx]: I thought about it at the beginning but I was not sure since the name studyUpdated refers to study modification. But yes it works fine.

File: src/components/computing-status/computing-type.ts (hunk 1 above)
Comment (AAJELLAL) [reply src/components/computing-status/computing-type.ts]: Done.

File: src/components/dialogs/parameters/sensi/sensitivity-analysis-parameters.jsx (hunk 4 above)
Comment (AAJELLAL) [reply src/components/dialogs/parameters/sensi/sensitivity-analysis-parameters.jsx:446]: Yes,no need for fetch. On the other hand for the change of **sensitivityAnalysisParams**, I prefer to do that in another US since this kind of refactoring is to be done for all parameters.

File: src/components/dialogs/parameters/short-circuit-parameters.jsx (hunk 5 above)
Comment (AAJELLAL) [reply src/components/dialogs/parameters/short-circuit-parameters.jsx]: Done.

File: src/components/map-viewer.jsx (hunk 8 above)
Comment (AAJELLAL) [reply src/components/map-viewer.jsx]: Done

File: src/redux/reducer.ts (hunk 9 above)
Comment (AAJELLAL) [reply src/redux/reducer.ts]: Yes, fixed.

File: src/redux/reducer.ts (hunk 10 above)
Comment (AAJELLAL) [reply src/redux/reducer.ts]: Done

File: src/components/dialogs/parameters/non-evacuated-energy/non-evacuated-energy-parameters.jsx [hunk 11]
```diff
@@ -81,8 +85,22 @@ export const useGetNonEvacuatedEnergyParameters = () => {
...
+
+    useEffect(() => {
+        if (studyUuid) {
+            fetchNonEvacuatedEnergyParameters(studyUuid);
+        }
+    }, [studyUuid, fetchNonEvacuatedEnergyParameters]);
+
+    // fetch the parameter if NON_EVACUATED_ENERGY_ANALYSIS  notification type is received.
+    useEffect(() => {
+        if (studyUuid && isComputationParametersUpdated(ComputationType.NON_EVACUATED_ENERGY_ANALYSIS, studyUpdated)) {
```
Comment (sBouzols) [inline src/components/dialogs/parameters/non-evacuated-energy/non-evacuated-energy-parameters.jsx:100]: Does NonEvacuatedEnergy computation needs to ckeck `optionalServiceStatus` ?

File: src/components/computing-status/computing-type.ts (hunk 1 above)
Comment (sBouzols) [reply src/components/computing-status/computing-type.ts]: This involve to much changes, Maybe to do in another PR

File: src/components/dialogs/parameters/parameters.jsx [hunk 12]
```diff
@@ -434,8 +414,27 @@ export const useParametersBackend = (
...
+        if (
+            isComputationParametersUpdated(type, studyUpdated) &&
+            backendFetchParameters &&
+            studyUuid &&
+            optionalServiceStatus === OptionalServicesStatus.Up
+        ) {
+            fetchParameters(studyUuid);
         }
-    }, [optionalServiceStatus, backendFetchParameters, snackError, studyUuid, type]);
+    }, [optionalServiceStatus, backendFetchParameters, fetchParameters, studyUuid, type, studyUpdated]);
```
[the remark this answers is not in the document]
Comment (sBouzols) [reply src/components/dialogs/parameters/parameters.jsx]: fixed using ref

File: src/components/dialogs/parameters/voltageinit/use-get-voltage-init-parameters.tsx [hunk 13]
```diff
@@ -36,8 +40,25 @@ export const useGetVoltageInitParameters = (): [
...
+    useEffect(() => {
+        if (
+            studyUuid &&
+            voltageInitAvailability === OptionalServicesStatus.Up &&
+            isComputationParametersUpdated(ComputationType.VOLTAGE_INITIALIZATION, studyUpdated)
+        ) {
+            fetchVoltageInitStudyParameters(studyUuid);
         }
-    }, [voltageInitAvailability, studyUuid, snackError]);
+    }, [studyUuid, voltageInitAvailability, fetchVoltageInitStudyParameters, studyUpdated]);
```
[the remark this answers is not in the document]
Comment (sBouzols) [reply src/components/dialogs/parameters/voltageinit/use-get-voltage-init-parameters.tsx]: fixed

File: src/components/dialogs/parameters/non-evacuated-energy/non-evacuated-energy-parameters.jsx (hunk 11 above)
Comment (sBouzols) [reply src/components/dialogs/parameters/non-evacuated-energy/non-evacuated-energy-parameters.jsx]: fixed

File: src/components/dialogs/parameters/shortcircuit/short-circuit-parameters-utils.tsx [hunk 14]
```diff
@@ -40,3 +51,21 @@ export const getStatus = (status: STATUS, styles: any) => {
     const color = status === STATUS.SUCCESS ? styles.succeed : styles.fail;
     return <Lens fontSize={'medium'} sx={color} />;
 };
+
+export const formatShortCircuitParameters = (
+    parameters: ShortCircuitParameters,
+    predefinedParameters: PREDEFINED_PARAMETERS
+): any => {
```
Comment (capyq) [inline src/components/dialogs/parameters/shortcircuit/short-circuit-parameters-utils.tsx:58]: RMQ => never use `any` type

Comment (sBouzols) [reply src/components/dialogs/parameters/shortcircuit/short-circuit-parameters-utils.tsx:58]: I didn't touch this line... i will put a TODO, there is an ongoing TS migration of those files.

Comment (capyq) [reply src/components/dialogs/parameters/shortcircuit/short-circuit-parameters-utils.tsx:58]: at least type this as Object

**Review by SlimaneAmar — APPROVED** [review]

**Review by sBouzols — APPROVED** [review]
Code Review OK
Tests done (multi user, multi tabs)
Console warning check OK
